### PR TITLE
1.x: Added Single execution hooks

### LINKED ISSUE #3595: Execution hook for Single
`RxJavaObservableExecutionHook` is not compatible with [`Single`](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/Single.java) type. For example: [`RxJavaObservableExecutionHook#onCreate`](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/plugins/RxJavaObservableExecutionHook.java#L54) method tooks [`Observable#OnSubscribe`](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/Observable.java#L101) type as a parameter. However, `Single` is using it's own [`OnSubscribe`](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/Single.java#L149) type. Shall we implement new class to handle `Single`'s hooks let's say `RxJavaSingleExecutionHook` ?

### PR DESCRIPTION
**Pull request by phajduk**
Resolves #3595

### CODE REVIEW

**Comment by akarnokd** [comment]
Would you also apply these plugin methods in Single?

  - onSubscribeStart
  - onSubscribeReturn
  - onSubscribeError
  - onLift

**Comment by phajduk** [comment]
Beloew hooks are already called (I replaced previous calls of `RxJavaObservableExecutionHook` by `RxJavaSingleExecutionHook` calls).
- onSubscribeReturn
- onSubscribeError
- onLift

However there is problem with `onSubscribeStart`. This hook should be called in `onSubscribe` and `unsafeSubscribe`. Unfortunately, in `Single` `onSubscribe` and `unsafeSubscribe` methods get `Subscriber` type as parameter instead of `SingleSubscriber`. I will make a commit to cover it soon.

**Comment by akarnokd** [comment]
There is something wrong, a bunch of Single test failed.

**Comment by akarnokd** [comment]
:+1:

File: src/main/java/rx/Single.java [hunk 1]
```diff
@@ -1569,14 +1569,12 @@ public final void onNext(T args) {
      * @param subscriber
      *            the Subscriber that will handle the emission or notification from the Single
      */
-    public final void unsafeSubscribe(Subscriber<? super T> subscriber) {
+    public final Subscription unsafeSubscribe(Subscriber<? super T> subscriber) {
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Looks like this is incompatible binary change https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.15

File: src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package rx.plugins;
+
+/**
+ * Default no-op implementation of {@link RxJavaSingleExecutionHook}
+ */
+/* package */class RxJavaSingleExecutionHookDefault extends RxJavaSingleExecutionHook {
+
+    private static RxJavaSingleExecutionHookDefault INSTANCE = new RxJavaSingleExecutionHookDefault();
```
Comment (artem-zinnatullin) [inline src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java]: `final`

File: src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package rx.plugins;
+
+/**
+ * Default no-op implementation of {@link RxJavaSingleExecutionHook}
+ */
+/* package */class RxJavaSingleExecutionHookDefault extends RxJavaSingleExecutionHook {
```
Comment (artem-zinnatullin) [inline src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java]: `/* package */` is kind of weird :) Let's remove it?

File: src/test/java/rx/plugins/RxJavaPluginsTest.java [hunk 4]
```diff
@@ -108,6 +108,14 @@ public void testObservableExecutionHookViaRegisterMethod() {
     }
 
     @Test
+    public void testSingleExecutionHookViaRegisterMethod() {
```
Comment (artem-zinnatullin) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: no need for `test` prefix

File: src/test/java/rx/plugins/RxJavaPluginsTest.java [hunk 5]
```diff
@@ -108,6 +108,14 @@ public void testObservableExecutionHookViaRegisterMethod() {
...
+        RxJavaPlugins p = new RxJavaPlugins();
+        p.registerSingleExecutionHook(new RxJavaSingleExecutionHookTestImpl());
+        RxJavaSingleExecutionHook impl = p.getSingleExecutionHook();
+        assertTrue(impl instanceof RxJavaSingleExecutionHookTestImpl);
```
Comment (artem-zinnatullin) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: Please change to `assertSame()`

File: src/test/java/rx/plugins/RxJavaPluginsTest.java [hunk 6]
```diff
@@ -238,6 +246,11 @@ public Date call(Date date) {
         // just use defaults
     }
 
+    // inside test so it is stripped from Javadocs
+    public static class RxJavaSingleExecutionHookTestImpl extends RxJavaSingleExecutionHook {
```
Comment (artem-zinnatullin) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: You can just use `Mockito.mock(RxJavaSingleExecutionHook.class)` for the test

File: src/main/java/rx/Single.java [hunk 7]
```diff
@@ -129,7 +129,7 @@ private Single(final Observable.OnSubscribe<T> f) {
      * @see <a href="http://reactivex.io/documentation/operators/create.html">ReactiveX operators documentation: Create</a>
      */
     public static <T> Single<T> create(OnSubscribe<T> f) {
-        return new Single<T>(f); // TODO need hook 
+        return new Single<T>(hook.onCreate(f));
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Needs test

File: src/main/java/rx/Single.java [hunk 8]
```diff
@@ -1569,14 +1569,12 @@ public final void onNext(T args) {
...
-    public final void unsafeSubscribe(Subscriber<? super T> subscriber) {
+    public final Subscription unsafeSubscribe(Subscriber<? super T> subscriber) {
         try {
             // new Subscriber so onStart it
             subscriber.onStart();
-            // TODO add back the hook
-            //            hook.onSubscribeStart(this, onSubscribe).call(subscriber);
-            onSubscribe.call(subscriber);
-            hook.onSubscribeReturn(subscriber);
+            hook.onSubscribeStart(this, onSubscribe).call(subscriber);
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: needs test

File: src/main/java/rx/Single.java [hunk 9]
```diff
@@ -1569,14 +1569,12 @@ public final void onNext(T args) {
...
+            return hook.onSubscribeReturn(subscriber);
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: needs test

File: src/main/java/rx/Single.java [hunk 10]
```diff
@@ -1593,6 +1591,7 @@ public final void unsafeSubscribe(Subscriber<? super T> subscriber) {
                 // TODO why aren't we throwing the hook's return value.
                 throw r;
             }
+            return Subscriptions.unsubscribed();
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: needs test

File: src/main/java/rx/Single.java [hunk 11]
```diff
@@ -1660,9 +1659,7 @@ public final Subscription subscribe(Subscriber<? super T> subscriber) {
         // The code below is exactly the same an unsafeSubscribe but not used because it would add a sigificent depth to alreay huge call stacks.
         try {
             // allow the hook to intercept and/or decorate
-            // TODO add back the hook
-            //            hook.onSubscribeStart(this, onSubscribe).call(subscriber);
-            onSubscribe.call(subscriber);
+            hook.onSubscribeStart(this, onSubscribe).call(subscriber);
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: needs test

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (phajduk) [inline src/main/java/rx/Single.java]: I strive to be as much as possible compatible with `Observable` type. Do you have some proposition how to handle it? If we remove `Subscription` as return type then we won't be able to decorate or replace the `Subscription` instance in hook.

File: src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java (hunk 2 above)
Comment (phajduk) [inline src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java]: I see similar issues in other places in RxJava codebase Shall I fix it in this PR or create separated one?

File: src/test/java/rx/plugins/RxJavaPluginsTest.java (hunk 4 above)
Comment (phajduk) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: Majority of test methods which we have are prefixed by `test`. [Contribution](https://github.com/ReactiveX/RxJava/blob/1.x/CONTRIBUTING.md) doc states to follow existing conventions. Do we have some separated guidelines for tests convention somewhere?

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Right right, I mean that we will have to release it as 1.2 instead of 1.1.x,  I guess. 

Maintainers: what do you think?

File: src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java (hunk 2 above)
Comment (artem-zinnatullin) [inline src/main/java/rx/plugins/RxJavaSingleExecutionHookDefault.java]: Please fix your case in this PR and feel free to submit separate one for other cases

File: src/test/java/rx/plugins/RxJavaPluginsTest.java (hunk 4 above)
Comment (artem-zinnatullin) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: Right, but `@Test` clearly shows that it's a test, I hope we'll have some kind of codestyle document, IDE settings and checkstyle rules in future

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/rx/Single.java]: Unsafesubscribe is called by operators and they don't use the returned Subscription. These unsubscribe the Subscriber.

Comment (JakeWharton) [inline src/main/java/rx/Single.java]: Single isn't stable API so this is acceptable if it makes the API better.

Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: k, let's just mention this as kind of warning in changelog

File: src/test/java/rx/plugins/RxJavaPluginsTest.java (hunk 4 above)
Comment (phajduk) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: If so then I believe it would be better to leave it as is and fix all test methods in bulk when we prepare IDE settings and SCA tools. What do you think?

Comment (artem-zinnatullin) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: k, as you wish, it doesn't really matter of course

File: src/test/java/rx/plugins/RxJavaPluginsTest.java (hunk 6 above)
Comment (phajduk) [inline src/test/java/rx/plugins/RxJavaPluginsTest.java]: I get your point. However I think `mock` won't help here as I need default behavior of default hook (otherwise in each test we will need to setup mock behavior). I would like to use `spy` to make both `verify` and default behavior possible. I'll make a commit soon so you'll be able to check my proposition.

**Comment by phajduk** [comment]
As far I see that failing test is not related with my changes. Is `rx.schedulers.CachedThreadSchedulerTest > testUnSubscribeForScheduler` some kind of flaky test?

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (stevegury) [inline src/main/java/rx/Single.java]: Yes this is a breaking binary change, but thanks to the fact that Single is not yet marked as stable, I vote for making this change.

File: src/main/java/rx/Single.java [hunk 12]
```diff
@@ -100,7 +100,7 @@ private Single(final Observable.OnSubscribe<T> f) {
         this.onSubscribe = f;
     }
 
-    static final RxJavaObservableExecutionHook hook = RxJavaPlugins.getInstance().getObservableExecutionHook();
+    static RxJavaSingleExecutionHook hook = RxJavaPlugins.getInstance().getSingleExecutionHook();
```
Comment (stevegury) [inline src/main/java/rx/Single.java]: Why did you remove `final`?

Comment (stevegury) [inline src/main/java/rx/Single.java]: Nevermind, I just saw your last comment.

**Comment by stevegury** [comment]
More tests would be appreciated, but otherwise LGTM.

**Comment by akarnokd** [comment]
Could you squash the commits and rebase this PR?

**Comment by stevegury** [comment]
:+1: (after you rebase/squash the commits)

**Comment by akarnokd** [comment]
:+1:
